### PR TITLE
Implement a custom fix all provider for SA1101

### PR DESCRIPTION
**Pull request by pdelvo**

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1101CodeFixProvider.cs [hunk 1]
```diff
@@ -67,5 +70,36 @@ private static Task<Document> GetTransformedDocumentAsync(Document document, Syn
 
             return Task.FromResult(document.WithSyntaxRoot(newSyntaxRoot));
         }
+
+        private class FixAll : DocumentBasedFixAllProvider
```
Comment (otac0n) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1101CodeFixProvider.cs]: The indentation for `Instance` and `FixAllInDocumentAsync { ... return ... }` is weird.

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/DocumentBasedFixAllProvider.cs [hunk 2]
```diff
@@ -42,21 +45,24 @@ internal abstract class DocumentBasedFixAllProvider : FixAllProvider
...
         private async Task<Solution> GetProjectFixesAsync(FixAllContext fixAllContext, Project project)
         {
             Solution solution = project.Solution;
-            var oldDocuments = project.Documents.ToImmutableArray();
-            List<Task<SyntaxNode>> newDocuments = new List<Task<SyntaxNode>>(oldDocuments.Length);
-            foreach (var document in oldDocuments)
+            List<Document> oldDocuments = new List<Document>();
+            List<Task<SyntaxNode>> newDocumentRoots = new List<Task<SyntaxNode>>();
+            var diagnostics = await fixAllContext.GetProjectDiagnosticsAsync(project).ConfigureAwait(false);
+            var lookup = diagnostics.GroupBy(x => project.GetDocument(x.Location.SourceTree)).ToArray();
```
Comment (otac0n) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/DocumentBasedFixAllProvider.cs]: Is `ToArray()` needed?  I don't know the internals of `project.GetDocument` but I would assume that avoiding an allocation here would be possible.

**Comment by pdelvo** [comment]
@otac0n I implemented your feedback

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/DocumentBasedFixAllProvider.cs [hunk 3]
```diff
@@ -42,21 +45,26 @@ internal abstract class DocumentBasedFixAllProvider : FixAllProvider
...
-                newDocuments.Add(this.FixAllInDocumentAsync(fixAllContext, document));
+                oldDocuments.Add(item.Key);
+                newDocumentRoots.Add(Task.Run(() => this.FixAllInDocumentAsync(fixAllContext, item.Key, item.ToImmutableArray()), fixAllContext.CancellationToken));
             }
 
-            for (int i = 0; i < oldDocuments.Length; i++)
+            for (int i = 0; i < oldDocuments.Count; i++)
             {
-                var newDocumentRoot = await newDocuments[i].ConfigureAwait(false);
+                fixAllContext.CancellationToken.ThrowIfCancellationRequested();
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/DocumentBasedFixAllProvider.cs]: :question: Is this actually necessary? It seems like the `await` on the following line would throw if cancellation occurred.

**Pull request merged**